src/storage/FirebaseStorage.js

```diff
@@ -0,0 +1,92 @@
+(function() {
+  "use strict";
+
+  function unwrapStorageSnapshot(storageSnapshot) {
+    return {
+      bytesTransferred: storageSnapshot.bytesTransferred,
+      downloadURL: storageSnapshot.downloadURL,
+      metadata: storageSnapshot.metadata,
+      ref: storageSnapshot.ref,
+      state: storageSnapshot.state,
+      task: storageSnapshot.task,
+      totalBytes: storageSnapshot.totalBytes
+    };
+  }
+
+  function isStorageRef(value) {
+    value = value || {};
+    return typeof value.put === 'function';
+  }
+
+  function _assertStorageRef(storageRef) {
+    if (!isStorageRef(storageRef)) {
+      throw new Error('$firebaseStorage expects a Storage reference');
+    }
+  }
+
+  function FirebaseStorage($firebaseUtils, $q) {
+
+    var Storage = function Storage(storageRef) {
+      _assertStorageRef(storageRef);
+      return {
+        $put: function $put(file) {
+          var task = storageRef.put(file);
+
+          return {
+            $progress: function $progress(callback) {
+              task.on('state_changed', function () {
+                $firebaseUtils.compile(function () {
+                  callback(unwrapStorageSnapshot(task.snapshot));
+                });
+              });
+            },
+            $error: function $error(callback) {
+              task.on('state_changed', null, function (err) {
+                $firebaseUtils.compile(function () {
+                  callback(err);
+                });
+              });
+            },
+            $complete: function $complete(callback) {
+              task.on('state_changed', null, null, function () {
+                $firebaseUtils.compile(function () {
+                  callback(unwrapStorageSnapshot(task.snapshot));
+                });
+              });
+            },
+            $cancel: task.cancel,
+            $resume: task.resume,
+            $pause: task.pause,
+            then: task.then,
+            catch: task.catch,
+            _task: task
```

Comment:
It looks like in the underlying SDK, `UploadTask` has a [`snapshot`](https://firebase.google.com/docs/reference/js/firebase.storage.UploadTask#snapshot) property. Since we are returning an object which is the AngularFire version of `UploadTask`, we should probably have `$snapshot: task.snapshot` here instead of `_task: task`.